File: viz-frontend/src/payload.ts
import type { VizData } from "./types";

/**
 * Payload transport between `fallow viz` and this bundle.
 *
 * The CLI writes the core payload into `<script id="fallow-data">` and each
 * large lens section into its own `<script data-fallow-lazy="path">` tag.
 * Both use `type="application/json"`, so the browser does not parse them as
 * script. The core is parsed at start. A lazy section is parsed on the first
 * read of its property, which is normally when its lens opens.
 *
 * Arrays of objects can arrive as tables: `{"$k": keys, "$r": rows}`. A
 * `null` cell or a missing trailing cell means that the key is absent.
 * The core tag has a `data-tables` attribute only when tables are in use.
 * The Rust side (`crates/cli/src/viz/payload.rs`) owns the encoder.
 */

/** One deferred payload section. `text` is called at most once. */
export interface LazySection {
  /** `a.b` puts the value at `data.a.b`; `a.*.b` spreads one entry per element of `data.a`. */
  path: string;
  text: () => string;
}

const TABLE_KEYS = "$k";
const TABLE_ROWS = "$r";
const COLUMN_MARKER = "*";

type JsonObject = Record<string, unknown>;

const isObject = (value: unknown): value is JsonObject =>
  typeof value === "object" && value !== null && !Array.isArray(value);

const isTable = (value: JsonObject): boolean => {
  const keys = Object.keys(value);
  return keys.length === 2 && Array.isArray(value[TABLE_KEYS]) && Array.isArray(value[TABLE_ROWS]);
};

const decodeRow = (keys: string[], row: unknown[]): JsonObject => {
  const out: JsonObject = {};
  for (let column = 0; column < keys.length && column < row.length; column++) {
    const cell = row[column];
    if (cell !== null) out[keys[column]] = decodeTables(cell);
  }
  return out;
};

/** Expand every `{"$k", "$r"}` table in a parsed JSON value into objects. */
export const decodeTables = (value: unknown): unknown => {
  if (Array.isArray(value)) return value.map(decodeTables);
  if (!isObject(value)) return value;
  if (isTable(value)) {
    const keys = value[TABLE_KEYS] as string[];
    return (value[TABLE_ROWS] as unknown[][]).map((row) => decodeRow(keys, row));
  }
  const out: JsonObject = {};
  for (const [key, entry] of Object.entries(value)) out[key] = decodeTables(entry);
  return out;
};

interface ParsedSection {
  path: string;
  parse: () => unknown;
}

const setValue = (target: JsonObject, key: string, value: unknown): void => {
  Object.defineProperty(target, key, {
    value,
    writable: true,
    enumerable: true,
    configurable: true,
  });
};

const defineLazy = (target: JsonObject, key: string, load: () => void): void => {
  Object.defineProperty(target, key, {
    configurable: true,
    enumerable: true,
    get: () => {
      load();
      return target[key];
    },
    set: (value: unknown) => setValue(target, key, value),
  });
};

const resolveParent = (root: JsonObject, segments: string[]): unknown => {
  let node: unknown = root;
  for (const segment of segments) {
    if (!isObject(node)) return undefined;
    node = node[segment];
  }
  return node;
};

const installProperty = (root: JsonObject, section: ParsedSection, segments: string[]): void => {
  const key = segments[segments.length - 1];
  const parent = resolveParent(root, segments.slice(0, -1));
  if (!isObject(parent)) return;
  defineLazy(parent, key, () => setValue(parent, key, section.parse()));
};

const installColumn = (root: JsonObject, section: ParsedSection, segments: string[]): void => {
  const marker = segments.indexOf(COLUMN_MARKER);
  const key = segments[marker + 1];
  const rows = resolveParent(root, segments.slice(0, marker));
  if (!Array.isArray(rows) || key === undefined) return;
  const elements = rows.filter(isObject);
  let loaded = false;
  const load = (): void => {
    if (loaded) return;
    loaded = true;
    let values: unknown[] = [];
    try {
      const column = section.parse();
      if (Array.isArray(column)) values = column;
    } finally {
      // Replace every getter, also when the parse fails, so a later read
      // cannot enter the getter again.
      elements.forEach((element, index) => {
        const value = values[index];
        if (Array.isArray(value) && value.length > 0) setValue(element, key, value);
        else delete element[key];
      });
    }
  };
  for (const element of elements) defineLazy(element, key, load);
};

/** Parse the core payload and attach each lazy section at its path. */
export const hydratePayload = (
  coreText: string,
  sections: LazySection[],
  tables = true,
): VizData => {
  const decode = (text: string): unknown =>
    tables ? decodeTables(JSON.parse(text)) : JSON.parse(text);
  const root = decode(coreText);
  if (!isObject(root)) throw new Error("fallow viz payload is not an object");
  for (const raw of sections) {
    const section: ParsedSection = { path: raw.path, parse: () => decode(raw.text()) };
    const segments = section.path.split(".");
    if (segments.includes(COLUMN_MARKER)) installColumn(root, section, segments);
    else installProperty(root, section, segments);
  }
  return root as unknown as VizData;
};

/** Read the payload tags that `fallow viz` writes into the page. */
export const readEmbeddedPayload = (doc: Document): VizData | null => {
  const core = doc.getElementById("fallow-data");
  const coreText = core?.textContent;
  if (!core || !coreText) return null;
  core.remove();
  const sections = [...doc.querySelectorAll<HTMLScriptElement>("script[data-fallow-lazy]")].map(
    (element): LazySection => ({
      path: element.dataset.fallowLazy ?? "",
      text: () => {
        // Drop the tag once it is read, so the page does not keep a
        // second copy of the section text in the DOM.
        const text = element.textContent ?? "null";
        element.remove();
        return text;
      },
    }),
  );
  return hydratePayload(coreText, sections, core.hasAttribute("data-tables"));
};

File: crates/cli/src/viz/payload.rs
//! Transport encoding for the payload that `fallow viz` embeds in its HTML.
//!
//! The payload has two parts:
//!
//! - The core. It holds what the first paint needs: files, edges, the
//!   summary and the availability of each analysis.
//! - Lazy sections. Each one holds a large finding list. The frontend parses
//!   a section on the first read of its property, which is normally when its
//!   lens opens.
//!
//! Arrays of objects become tables: `{"$k": keys, "$r": rows}`. A `null`
//! cell or a missing trailing cell means that the key is absent. The encoder
//! keeps an array plain when the table is not shorter or when a member holds
//! an explicit `null`. When any key in the payload starts with `$`, the
//! encoder writes no tables at all and `tables` is false, so the page tells
//! the decoder to leave such objects as they are.
//!
//! `viz-frontend/src/payload.ts` owns the decoder. Both sides pin the same
//! fixture, so they cannot drift apart.

use rustc_hash::FxHashSet;
use serde_json::{Map, Value};

use fallow_engine::viz::VizData;

/// Sections that the first paint does not read. Each path names an object
/// key, with `.` between the levels.
const LAZY_SECTIONS: &[&str] = &[
    "health.files",
    "health.findings",
    "security.candidates",
    "security.blind_spots",
    "styling.findings",
    "dependencies.findings",
    "architecture.findings",
    "frameworks.findings",
    "feature_flags.findings",
    "clones",
];

/// The per-file function lists, sent as one array that aligns with `files`.
const FUNCTIONS_COLUMN: &str = "files.*.functions";

const TABLE_KEYS: &str = "$k";
const TABLE_ROWS: &str = "$r";
const RESERVED_PREFIX: char = '$';

/// The encoded payload: the core JSON and one JSON text per lazy section.
pub(super) struct EncodedPayload {
    pub core: String,
    pub lazy: Vec<(&'static str, String)>,
    /// Whether the texts can hold tables that the decoder must expand.
    pub tables: bool,
}

/// Split and encode the viz payload for the HTML page.
pub(super) fn encode_payload(data: &VizData) -> Result<EncodedPayload, serde_json::Error> {
    encode_value(serde_json::to_value(data)?)
}

fn encode_value(mut root: Value) -> Result<EncodedPayload, serde_json::Error> {
    let tables = !has_reserved_key(&root);
    let finish = |value: Value| {
        let value = if tables { encode_tables(value) } else { value };
        serde_json::to_string(&value)
    };

    let mut lazy = Vec::new();
    if let Some(column) = take_column(&mut root, "files", "functions") {
        lazy.push((FUNCTIONS_COLUMN, finish(column)?));
    }
    for path in LAZY_SECTIONS {
        if let Some(section) = take_path(&mut root, path) {
            lazy.push((*path, finish(section)?));
        }
    }
    Ok(EncodedPayload {
        core: finish(root)?,
        lazy,
        tables,
    })
}

/// Remove the value at a dotted object path and return it.
fn take_path(root: &mut Value, path: &str) -> Option<Value> {
    let (parent_path, key) = path.rsplit_once('.').unwrap_or(("", path));
    let mut parent = root;
    for segment in parent_path.split('.').filter(|segment| !segment.is_empty()) {
        parent = parent.get_mut(segment)?;
    }
    parent.as_object_mut()?.remove(key)
}

/// Remove `key` from each object in the `rows` array. Return one entry per
/// row (an empty array when the row had no value), or `None` when no row had
/// a value.
fn take_column(root: &mut Value, rows: &str, key: &str) -> Option<Value> {
    let rows = root.get_mut(rows)?.as_array_mut()?;
    let mut found = false;
    let column: Vec<Value> = rows
        .iter_mut()
        .map(|row| {
            let value = row.as_object_mut().and_then(|row| row.remove(key));
            found |= value.is_some();
            value.unwrap_or_else(|| Value::Array(Vec::new()))
        })
        .collect();
    found.then_some(Value::Array(column))
}

fn has_reserved_key(value: &Value) -> bool {
    match value {
        Value::Object(map) => map
            .iter()
            .any(|(key, value)| key.starts_with(RESERVED_PREFIX) || has_reserved_key(value)),
        Value::Array(items) => items.iter().any(has_reserved_key),
        _ => false,
    }
}

/// Encode each array of objects as a table when the table is shorter.
fn encode_tables(value: Value) -> Value {
    match value {
        Value::Object(map) => Value::Object(
            map.into_iter()
                .map(|(key, value)| (key, encode_tables(value)))
                .collect(),
        ),
        Value::Array(items) => {
            let items: Vec<Value> = items.into_iter().map(encode_tables).collect();
            match table_of(&items) {
                Some(table) if json_len(&table) < json_len_of_items(&items) => table,
                _ => Value::Array(items),
            }
        }
        other => other,
    }
}

fn table_of(items: &[Value]) -> Option<Value> {
    if items.len() < 2 {
        return None;
    }
    let mut keys: Vec<&str> = Vec::new();
    let mut seen: FxHashSet<&str> = FxHashSet::default();
    for item in items {
        let object = item.as_object()?;
        for (key, value) in object {
            // `null` marks an absent key in a row, so a real `null` cannot
            // travel in a table.
            if value.is_null() {
                return None;
            }
            if seen.insert(key.as_str()) {
                keys.push(key.as_str());
            }
        }
    }
    let rows = items
        .iter()
        .filter_map(Value::as_object)
        .map(|object| Value::Array(table_row(object, &keys)))
        .collect();
    let mut table = Map::new();
    table.insert(
        TABLE_KEYS.to_string(),
        Value::Array(
            keys.iter()
                .map(|key| Value::String((*key).to_string()))
                .collect(),
        ),
    );
    table.insert(TABLE_ROWS.to_string(), Value::Array(rows));
    Some(Value::Object(table))
}

fn table_row(object: &Map<String, Value>, keys: &[&str]) -> Vec<Value> {
    let mut row: Vec<Value> = keys
        .iter()
        .map(|key| object.get(*key).cloned().unwrap_or(Value::Null))
        .collect();
    while row.last().is_some_and(Value::is_null) {
        row.pop();
    }
    row
}

fn json_len(value: &Value) -> usize {
    serde_json::to_string(value).map_or(usize::MAX, |text| text.len())
}

fn json_len_of_items(items: &[Value]) -> usize {
    // `[` + items + `,` separators + `]`.
    items.iter().map(json_len).sum::<usize>() + items.len().saturating_sub(1) + 2
}

#[cfg(test)]
mod tests {
    use serde_json::json;

    use super::*;

    /// Mirror of `decodeTables` in `viz-frontend/src/payload.ts`.
    fn decode_tables(value: Value) -> Value {
        match value {
            Value::Array(items) => Value::Array(items.into_iter().map(decode_tables).collect()),
            Value::Object(mut map)
                if map.len() == 2
                    && map.get(TABLE_KEYS).is_some_and(Value::is_array)
                    && map.get(TABLE_ROWS).is_some_and(Value::is_array) =>
            {
                let Some(Value::Array(keys)) = map.remove(TABLE_KEYS) else {
                    unreachable!("checked above");
                };
                let Some(Value::Array(rows)) = map.remove(TABLE_ROWS) else {
                    unreachable!("checked above");
                };
                let rows = rows
                    .into_iter()
                    .map(|row| {
                        let Value::Array(cells) = row else {
                            panic!("table row is not an array");
                        };
                        let object: Map<String, Value> = keys
                            .iter()
                            .zip(cells)
                            .filter(|(_, cell)| !cell.is_null())
                            .map(|(key, cell)| {
                                (key.as_str().expect("key").to_string(), decode_tables(cell))
                            })
                            .collect();
                        Value::Object(object)
                    })
                    .collect();
                Value::Array(rows)
            }
            Value::Object(map) => Value::Object(
                map.into_iter()
                    .map(|(key, value)| (key, decode_tables(value)))
                    .collect(),
            ),
            other => other,
        }
    }

    /// Rebuild the full payload the way the frontend does after every lazy
    /// section has been read.
    fn hydrate(encoded: &EncodedPayload) -> Value {
        let decode = |text: &str| {
            let value: Value = serde_json::from_str(text).expect("payload JSON");
            if encoded.tables {
                decode_tables(value)
            } else {
                value
            }
        };
        let mut root = decode(&encoded.core);
        for (path, text) in &encoded.lazy {
            let section = decode(text);
            if *path == FUNCTIONS_COLUMN {
                let Value::Array(column) = section else {
                    panic!("function column is not an array");
                };
                let files = root["files"].as_array_mut().expect("files");
                for (file, functions) in files.iter_mut().zip(column) {
                    if functions.as_array().is_some_and(|list| !list.is_empty()) {
                        file.as_object_mut()
                            .expect("file")
                            .insert("functions".to_string(), functions);
                    }
                }
                continue;
            }
            let (parent_path, key) = path.rsplit_once('.').unwrap_or(("", path));
            let mut parent = &mut root;
            for segment in parent_path.split('.').filter(|segment| !segment.is_empty()) {
                parent = parent.get_mut(segment).expect("parent");
            }
            parent
                .as_object_mut()
                .expect("parent object")
                .insert((*key).to_string(), section);
        }
        root
    }

    fn sample() -> Value {
        json!({
            "root": "demo",
            "files": [
                {"path": "a.ts", "size": 10, "functions": [
                    {"name": "f", "line": 1, "cyclomatic": 2},
                    {"name": "g", "line": 9, "cyclomatic": 1},
                    {"name": "h", "line": 20, "cyclomatic": 4}
                ]},
                {"path": "b.ts", "size": 20, "zone": 1},
                {"path": "c.ts", "size": 30}
            ],
            "edges": [[0, 1, 0], [1, 2, 1]],
            "clones": [],
            "health": {
                "availability": {"state": "complete", "count": 2, "unit": "files"},
                "score": 90.5,
                "files": [],
                "findings": [
                    {"kind": "health-finding", "file": 0, "facts": [
                        {"label": "cyclomatic", "value": "30"},
                        {"label": "cognitive", "value": "40"}
                    ]},
                    {"kind": "health-finding", "file": 1, "severity": "high"}
                ]
            },
            "styling": {"availability": {"state": "disabled", "count": 0, "unit": "findings"},
                        "findings": [], "summary": {"tokens": [1, 2]}}
        })
    }

    #[test]
    fn encoded_table_matches_the_frontend_fixture() {
        // `viz-frontend/src/payload.test.ts` decodes this exact text.
        let rows = json!([
            {"line": 1, "name": "a", "props": 2},
            {"line": 2, "name": "b"},
            {"line": 3, "props": 4}
        ]);
        assert_eq!(
            serde_json::to_string(&encode_tables(rows)).expect("serialize"),
            r#"{"$k":["line","name","props"],"$r":[[1,"a",2],[2,"b"],[3,null,4]]}"#
        );
    }

    #[test]
    fn hydrated_payload_equals_the_original() {
        let original = sample();
        let encoded = encode_value(original.clone()).expect("encode");
        assert_eq!(hydrate(&encoded), original);
    }

    #[test]
    fn core_leaves_out_every_lazy_section() {
        let encoded = encode_value(sample()).expect("encode");
        let core: Value = serde_json::from_str(&encoded.core).expect("core JSON");
        assert!(core["health"].get("findings").is_none());
        assert!(core["health"].get("files").is_none());
        assert!(core.get("clones").is_none());
        assert_eq!(core["health"]["score"], json!(90.5));
        assert!(!encoded.core.contains("\"functions\""));
        let paths: Vec<&str> = encoded.lazy.iter().map(|(path, _)| *path).collect();
        assert_eq!(
            paths,
            [
                FUNCTIONS_COLUMN,
                "health.files",
                "health.findings",
                "styling.findings",
                "clones"
            ]
        );
    }

    #[test]
    fn function_column_is_absent_when_no_file_has_functions() {
        let mut value = sample();
        value["files"][0]
            .as_object_mut()
            .expect("file")
            .remove("functions");
        let encoded = encode_value(value.clone()).expect("encode");
        assert!(
            encoded
                .lazy
                .iter()
                .all(|(path, _)| *path != FUNCTIONS_COLUMN)
        );
        assert_eq!(hydrate(&encoded), value);
    }

    #[test]
    fn table_is_not_used_when_it_is_longer() {
        let rows = json!([{"a": 1}, {"a": 2}]);
        assert_eq!(encode_tables(rows.clone()), rows);
    }

    #[test]
    fn arrays_with_explicit_null_stay_plain() {
        let rows = json!([
            {"name": "alpha", "value": null},
            {"name": "beta", "value": 1},
            {"name": "gamma", "value": 2}
        ]);
        assert_eq!(encode_tables(rows.clone()), rows);
    }

    #[test]
    fn reserved_key_turns_table_encoding_off() {
        let mut value = sample();
        value["styling"]["summary"] = json!({"$k": ["x"], "$r": [[1]]});
        let encoded = encode_value(value.clone()).expect("encode");
        assert!(!encoded.tables);
        let core: Value = serde_json::from_str(&encoded.core).expect("core JSON");
        assert!(core["files"].is_array());
        assert_eq!(hydrate(&encoded), value);
    }
}
